# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/OriginOperations.cpp
# Commit: 46a5babaafd2
# Full Hash: 46a5babaafd2c0f2198f353d62254cf518dab391
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-15 04:16:11
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Implement initialization of entire persistent storage; r=dom-storage-reviewers,jari
#   
#   IndexedDB still calls QuotaManager::EnsurePersistentOriginIsInitializedInternal
#   from its idle database maintenance. However, we want to get rid of that so the
#   method can be made private. The idle database maintenance should only rely on
# ==============================================================================

diff -r 2ff5f3aaf841 -r 46a5babaafd2 dom/quota/OriginOperations.cpp
--- a/dom/quota/OriginOperations.cpp	Mon Oct 14 10:00:26 2024 +0000
+++ b/dom/quota/OriginOperations.cpp	Mon Oct 14 10:06:14 2024 +0000
@@ -337,12 +337,27 @@
   bool GetResolveValue() override;
 };
 
+class PersistentStorageInitializedOp final : public InitializedRequestBase {
+ public:
+  explicit PersistentStorageInitializedOp(
+      MovingNotNull<RefPtr<QuotaManager>> aQuotaManager)
+      : InitializedRequestBase(std::move(aQuotaManager),
+                               "dom::quota::PersistentStorageInitializedOp") {}
+
+ private:
+  ~PersistentStorageInitializedOp() = default;
+
+  nsresult DoDirectoryWork(QuotaManager& aQuotaManager) override;
+
+  bool GetResolveValue() override;
+};
+
 class TemporaryStorageInitializedOp final : public InitializedRequestBase {
  public:
   explicit TemporaryStorageInitializedOp(
       MovingNotNull<RefPtr<QuotaManager>> aQuotaManager)
       : InitializedRequestBase(std::move(aQuotaManager),
-                               "dom::quota::StorageInitializedOp") {}
+                               "dom::quota::TemporaryStorageInitializedOp") {}
 
  private:
   ~TemporaryStorageInitializedOp() = default;
@@ -420,6 +435,27 @@
   void CloseDirectory() override;
 };
 
+class InitializePersistentStorageOp final
+    : public ResolvableNormalOriginOp<bool> {
+  RefPtr<UniversalDirectoryLock> mDirectoryLock;
+
+ public:
+  InitializePersistentStorageOp(
+      MovingNotNull<RefPtr<QuotaManager>> aQuotaManager,
+      RefPtr<UniversalDirectoryLock> aDirectoryLock);
+
+ private:
+  ~InitializePersistentStorageOp() = default;
+
+  RefPtr<BoolPromise> OpenDirectory() override;
+
+  nsresult DoDirectoryWork(QuotaManager& aQuotaManager) override;
+
+  bool GetResolveValue() override;
+
+  void CloseDirectory() override;
+};
+
 class InitTemporaryStorageOp final : public ResolvableNormalOriginOp<bool> {
   RefPtr<UniversalDirectoryLock> mDirectoryLock;
 
@@ -941,6 +977,11 @@
   return MakeRefPtr<StorageInitializedOp>(std::move(aQuotaManager));
 }
 
+RefPtr<ResolvableNormalOriginOp<bool>> CreatePersistentStorageInitializedOp(
+    MovingNotNull<RefPtr<QuotaManager>> aQuotaManager) {
+  return MakeRefPtr<PersistentStorageInitializedOp>(std::move(aQuotaManager));
+}
+
 RefPtr<ResolvableNormalOriginOp<bool>> CreateTemporaryStorageInitializedOp(
     MovingNotNull<RefPtr<QuotaManager>> aQuotaManager) {
   return MakeRefPtr<TemporaryStorageInitializedOp>(std::move(aQuotaManager));
@@ -968,6 +1009,13 @@
                             std::move(aDirectoryLock));
 }
 
+RefPtr<ResolvableNormalOriginOp<bool>> CreateInitializePersistentStorageOp(
+    MovingNotNull<RefPtr<QuotaManager>> aQuotaManager,
+    RefPtr<UniversalDirectoryLock> aDirectoryLock) {
+  return MakeRefPtr<InitializePersistentStorageOp>(std::move(aQuotaManager),
+                                                   std::move(aDirectoryLock));
+}
+
 RefPtr<ResolvableNormalOriginOp<bool>> CreateInitTemporaryStorageOp(
     MovingNotNull<RefPtr<QuotaManager>> aQuotaManager,
     RefPtr<UniversalDirectoryLock> aDirectoryLock) {
@@ -1736,6 +1784,23 @@
   return mInitialized;
 }
 
+nsresult PersistentStorageInitializedOp::DoDirectoryWork(
+    QuotaManager& aQuotaManager) {
+  AssertIsOnIOThread();
+
+  AUTO_PROFILER_LABEL("PersistentStorageInitializedOp::DoDirectoryWork", OTHER);
+
+  mInitialized = aQuotaManager.IsPersistentStorageInitializedInternal();
+
+  return NS_OK;
+}
+
+bool PersistentStorageInitializedOp::GetResolveValue() {
+  AssertIsOnOwningThread();
+
+  return mInitialized;
+}
+
 nsresult TemporaryStorageInitializedOp::DoDirectoryWork(
     QuotaManager& aQuotaManager) {
   AssertIsOnIOThread();
@@ -1872,6 +1937,49 @@
   DropDirectoryLock(mDirectoryLock);
 }
 
+InitializePersistentStorageOp::InitializePersistentStorageOp(
+    MovingNotNull<RefPtr<QuotaManager>> aQuotaManager,
+    RefPtr<UniversalDirectoryLock> aDirectoryLock)
+    : ResolvableNormalOriginOp(std::move(aQuotaManager),
+                               "dom::quota::InitializePersistentStorageOp"),
+      mDirectoryLock(std::move(aDirectoryLock)) {
+  AssertIsOnOwningThread();
+}
+
+RefPtr<BoolPromise> InitializePersistentStorageOp::OpenDirectory() {
+  AssertIsOnOwningThread();
+  MOZ_ASSERT(mDirectoryLock);
+
+  return BoolPromise::CreateAndResolve(true, __func__);
+}
+
+nsresult InitializePersistentStorageOp::DoDirectoryWork(
+    QuotaManager& aQuotaManager) {
+  AssertIsOnIOThread();
+
+  AUTO_PROFILER_LABEL("InitializePersistentStorageOp::DoDirectoryWork", OTHER);
+
+  QM_TRY(OkIf(aQuotaManager.IsStorageInitializedInternal()),
+         NS_ERROR_NOT_INITIALIZED);
+
+  QM_TRY(MOZ_TO_RESULT(
+      aQuotaManager.EnsurePersistentStorageIsInitializedInternal()));
+
+  return NS_OK;
+}
+
+bool InitializePersistentStorageOp::GetResolveValue() {
+  AssertIsOnOwningThread();
+
+  return true;
+}
+
+void InitializePersistentStorageOp::CloseDirectory() {
+  AssertIsOnOwningThread();
+
+  DropDirectoryLock(mDirectoryLock);
+}
+
 InitTemporaryStorageOp::InitTemporaryStorageOp(
     MovingNotNull<RefPtr<QuotaManager>> aQuotaManager,
     RefPtr<UniversalDirectoryLock> aDirectoryLock)